service/tracker/execution_test.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package tracker_test
+
+import (
+	"testing"
+
+	"github.com/gammazero/deque"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/engine/execution/computation/computer/uploader"
+
+	"github.com/optakt/flow-dps/service/tracker"
+	"github.com/optakt/flow-dps/testing/mocks"
+)
+
+func TestExecution_Update(t *testing.T) {
+	blockData := mocks.GenericBlockData()
+
+	t.Run("nominal case with nothing in queue", func(t *testing.T) {
+		t.Parallel()
+
+		streamer := mocks.BaselineRecordStreamer(t)
+		streamer.NextFunc = func() (*uploader.BlockData, error) {
+			return blockData, nil
+		}
+
+		exec := tracker.BaselineExecution(t, tracker.WithStreamer(streamer))
+
+		got, err := exec.Update()
+
+		require.NoError(t, err)
+		assert.Equal(t, blockData.TrieUpdates[0], got)
+	})
+
+	t.Run("nominal case with queue already filled", func(t *testing.T) {
+		t.Parallel()
+
+		streamer := mocks.BaselineRecordStreamer(t)
+		streamer.NextFunc = func() (*uploader.BlockData, error) {
+			t.Fail() // This should never be called, since the queue is already filled.
+			return nil, nil
+		}
+
+		queue := deque.New()
+		queue.PushBack(blockData.TrieUpdates[0])
```

Comment:
This is a bad test case, because you use data that is found in another place within the component. It won't allow you to be unambiguously sure about where the data comes from without making an assumption. I know you put the `t.Fail()`, but that assumes something about the internal logic then. We should have a different trie update here.

Reply:
I changed it to the following:

```go
		streamer := mocks.BaselineRecordStreamer(t)
		streamer.NextFunc = func() (*uploader.BlockData, error) {
			t.Fatal("unexpected call to streamer.Next()") // This should never be called, since the queue is already filled.
			return record, nil
		}

		queue := deque.New()
		queue.PushBack(update)
```

This way `update` and `record.TrieUpdates[0]` can't be the same, so it would not only error because of the `t.Fatal` if we were to use the return value of the `Next` call, but also because it would fetch the wrong update.

SGTY?

Reply:
I'm not sure why you changed the fail, but yes, just using a different update here solves the implicit dependency.